File: clojure/all-your-base/src/all_your_base.clj
(ns all-your-base)

(defn exp [x n]
  (reduce * (repeat n x)))

(defn factors [num base]
  (reverse (for [digit (range)
                 :let [factor (exp base digit)]
                 :while (<= factor num)] factor)))

(defn digits [num factors]
  (cond
    (empty? factors) []
    :else (let [digit (unchecked-divide-int num (first factors))
                remainder (mod num (first factors))]
            (cons digit (digits remainder (rest factors))))))

(defn dec-to-base [num base]
  (digits num (factors num base)))

(defn base-to-dec [base num] (reduce #(+ %2 (* %1 base)) 0 num))

(defn convert [a num b]
  (cond
    (or (< a 2) (< b 2)) nil
    (empty? num) []
    (some #(or (>= % a) (< % 0)) num) nil
    (every? #(= % 0) num) [0] 
    :else (dec-to-base (base-to-dec a num) b)))

File: lua/all-your-base/all-your-base.lua
local all_your_base = {}

function reverse(tab)
  for i = 1, #tab // 2, 1 do
    tab[i], tab[#tab - i + 1] = tab[#tab - i + 1], tab[i]
  end
  return tab
end


function base_to_dec(base, digits)
  local dec = 0
  for i = 1, #digits do
    assert(digits[i] >= 0, 'negative digits are not allowed')
    assert(digits[i] < base, 'digit out of range')
    dec = dec * base + digits[i]
  end
  return dec
end

function dec_to_base(num, base)
  local digits = {}
  while num > 0 do
    local digit = num % base
    table.insert(digits, digit)
    num = (num - digit) / base
  end
  return reverse(digits)
end

all_your_base.convert = function(from_digits, from_base)
  return {
    to = function (to_base)
      assert(from_base >= 2, 'invalid input base')
      assert(to_base >= 2, 'invalid output base')
      local dec = base_to_dec(from_base, from_digits)
      if dec == 0 then
        return { 0 }
      else
        return dec_to_base(dec, to_base)
      end
    end
  }
end

return all_your_base
